src/util/timer.h

```diff
@@ -33,103 +35,58 @@ class Timer {
-
-  private:
-    mixxx::Duration m_leapTime;
-};
-
 class ScopedTimer {
   public:
-    ScopedTimer(const char* key, int i,
-                Stat::ComputeFlags compute = kDefaultComputeFlags)
-            : m_pTimer(NULL),
-              m_cancel(false) {
-        if (CmdlineArgs::Instance().getDeveloper()) {
-            initialize(QString(key), QString::number(i), compute);
-        }
+    ScopedTimer(QStringView key,
+            Stat::ComputeFlags compute = kDefaultComputeFlags)
+            : ScopedTimer(key, QStringView(), compute) {
     }
-
-    ScopedTimer(const char* key, const char *arg = NULL,
-                Stat::ComputeFlags compute = kDefaultComputeFlags)
-            : m_pTimer(NULL),
-              m_cancel(false) {
-        if (CmdlineArgs::Instance().getDeveloper()) {
-            initialize(QString(key), arg ? QString(arg) : QString(), compute);
-        }
+    ScopedTimer(QStringView key, int i, Stat::ComputeFlags compute = kDefaultComputeFlags)
+            : ScopedTimer(key, QString::number(i), compute) {
     }
 
-    ScopedTimer(const char* key, const QString& arg,
-                Stat::ComputeFlags compute = kDefaultComputeFlags)
-            : m_pTimer(NULL),
-              m_cancel(false) {
-        if (CmdlineArgs::Instance().getDeveloper()) {
-            initialize(QString(key), arg, compute);
+    ScopedTimer(QStringView key, QStringView arg, Stat::ComputeFlags compute = kDefaultComputeFlags)
+            : m_state(std::nullopt) {
+        if (!CmdlineArgs::Instance().getDeveloper()) {
+            return;
         }
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+        QString strKey = arg.isEmpty() ? key.toString() : key.arg(arg);
+#else
+        QString strKey = arg.isEmpty() ? key.toString() : key.toString().arg(arg);
+#endif
+        m_state = std::make_optional<TimerData>(TimerData{Timer(strKey, compute), false});
+        m_state->timer.start();
     }
 
-    virtual ~ScopedTimer() {
-        if (m_pTimer) {
-            if (!m_cancel) {
-                m_pTimer->elapsed(true);
+    ~ScopedTimer() noexcept {
+        if (m_state) {
+            if (!m_state->cancelled) {
+                m_state->timer.elapsed(true);
             }
-            m_pTimer->~Timer();
         }
     }
 
-    inline void initialize(const QString& key, const QString& arg,
-                Stat::ComputeFlags compute = kDefaultComputeFlags) {
-        QString strKey;
-        if (arg.isEmpty()) {
-            strKey = key;
-        } else {
-            strKey = key.arg(arg);
-        }
-        m_pTimer = new(m_timerMem) Timer(strKey, compute);
-        m_pTimer->start();
-    }
+    ScopedTimer(const ScopedTimer&) = delete;
+    ScopedTimer& operator=(const ScopedTimer&) = delete;
+
+    ScopedTimer(ScopedTimer&&) = default;
+    ScopedTimer& operator=(ScopedTimer&&) = default;
 
     void cancel() {
-        m_cancel = true;
+        if (m_state) {
+            m_state->cancelled = true;
+        }
```

Comment:
The cancelled state can be replaced by m_state == std::nullopt;
 ```suggestion
        m_state.reset();
```
Because the timers is no longer required and can be destructed.

Reply:
I'm not super familiar with C++ destructors but I thought this would still happen automatically for RAII types even when using a user-provided constructor. Or is there a way to call a default-generated destructor within the user-provided destructor?

Reply:
> The cancelled state can be replaced by m_state == std::nullopt;

That implies the removal of the `cancel` method, but its only used in one location and thats before a return, so basically useless anyway.